src/db.c

```diff
@@ -1914,98 +1913,93 @@ int xreadGetKeys(struct redisCommand *cmd, robj **argv, int argc, getKeysResult
     return num;
 }
 
+/* In cluster mode, links to the next and previous entries in the same slot are
+ * stored in the dict entry metadata. */
+#define dictEntryNextInSlot(de) (((dictEntry **)dictMetadata(de))[0])
+#define dictEntryPrevInSlot(de) (((dictEntry **)dictMetadata(de))[1])
+
 /* Slot to Key API. This is used by Redis Cluster in order to obtain in
  * a fast way a key that belongs to a specified hash slot. This is useful
  * while rehashing the cluster and in other conditions when we need to
  * understand if we have keys for a given hash slot. */
-void slotToKeyUpdateKey(sds key, int add) {
+void slotToKeyUpdateKey(dictEntry *entry, int add) {
+    sds key = entry->key;
     size_t keylen = sdslen(key);
     unsigned int hashslot = keyHashSlot(key,keylen);
-    unsigned char buf[64];
-    unsigned char *indexed = buf;
 
     server.cluster->slots_keys_count[hashslot] += add ? 1 : -1;
-    if (keylen+2 > 64) indexed = zmalloc(keylen+2);
-    indexed[0] = (hashslot >> 8) & 0xff;
-    indexed[1] = hashslot & 0xff;
-    memcpy(indexed+2,key,keylen);
     if (add) {
-        raxInsert(server.cluster->slots_to_keys,indexed,keylen+2,NULL,NULL);
+        /* Add entry before the first element in the list. */
+        dictEntry *first = server.cluster->slots_to_keys[hashslot];
+        dictEntryNextInSlot(entry) = first;
+        if (first != NULL) {
+            serverAssert(dictEntryPrevInSlot(first) == NULL);
+            dictEntryPrevInSlot(first) = entry;
+        }
+        serverAssert(dictEntryPrevInSlot(entry) == NULL);
+        server.cluster->slots_to_keys[hashslot] = entry;
     } else {
-        raxRemove(server.cluster->slots_to_keys,indexed,keylen+2,NULL);
+        /* Remove entry from the linked list and connect the prev and next
+         * entries. */
+        dictEntry *next = dictEntryNextInSlot(entry);
+        dictEntry *prev = dictEntryPrevInSlot(entry);
+        if (next != NULL) {
+            dictEntryPrevInSlot(next) = prev;
+        }
+        if (prev != NULL) {
+            dictEntryNextInSlot(prev) = next;
+        } else {
+            /* The removed entry was the first in the list. */
+            serverAssert(server.cluster->slots_to_keys[hashslot] == entry);
+            server.cluster->slots_to_keys[hashslot] = next;
+        }
     }
-    if (indexed != buf) zfree(indexed);
 }
 
-void slotToKeyAdd(sds key) {
-    slotToKeyUpdateKey(key,1);
+void slotToKeyAdd(dictEntry *de) {
```

Comment:
All of these function names are now a bit wrong since you are passing in a dictEntry and not a key (Which is kind of key). I wonder if maybe we should make an alias like `keyHandle` that we can use instead of dictEntry to improve clarity?

Reply:
Good point. Aliasing the dictEntry as keyHandle (typedef?) can be confusing though. Maybe we can rename the function to slotToEntryAdd?